pkg/controller/volume/populator/desired_state_of_world_populator.go

```diff
@@ -0,0 +1,92 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package populator implements interfaces that monitor and keep the states of the
+// desired_state_of_word in sync with the "ground truth" from informer.
+
+package populator
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	kcache "k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/controller/volume/cache"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
+)
+
+// DesiredStateOfWorldPopulator periodically verifies that the pods in the
+// desired state of th world still exist, if not, it removes them.
+// TODO: it also loops through the list of active pods and ensures that
+// each one exists in the desired state of the world cache
+// if it has volumes.
+type DesiredStateOfWorldPopulator interface {
+	Run(stopCh <-chan struct{})
+}
+
+// NewDesiredStateOfWorldPopulator returns a new instance of DesiredStateOfWorldPopulator.
+// loopSleepDuration - the amount of time the populator loop sleeps between
+//     successive executions
+// podManager - the kubelet podManager that is the source of truth for the pods
+//     that exist on this host
+// desiredStateOfWorld - the cache to populate
+func NewDesiredStateOfWorldPopulator(
+	loopSleepDuration time.Duration,
+	podInformer framework.SharedInformer,
+	desiredStateOfWorld cache.DesiredStateOfWorld) DesiredStateOfWorldPopulator {
+	return &desiredStateOfWorldPopulator{
+		loopSleepDuration:   loopSleepDuration,
+		podInformer:         podInformer,
+		desiredStateOfWorld: desiredStateOfWorld,
+	}
+}
+
+type desiredStateOfWorldPopulator struct {
+	loopSleepDuration   time.Duration
+	podInformer         framework.SharedInformer
+	desiredStateOfWorld cache.DesiredStateOfWorld
+}
+
+func (dswp *desiredStateOfWorldPopulator) Run(stopCh <-chan struct{}) {
+	wait.Until(dswp.populatorLoopFunc(), dswp.loopSleepDuration, stopCh)
+}
+
+func (dswp *desiredStateOfWorldPopulator) populatorLoopFunc() func() {
+	return func() {
+		dswp.findAndRemoveDeletedPods()
+	}
+}
+
+// Iterate through all pods in desired state of world, and remove if they no
+// longer exist in the informer
+func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
+	for podUID, pod := range dswp.desiredStateOfWorld.GetPods() {
+		podKey, err := kcache.MetaNamespaceKeyFunc(pod.Pod)
+		podObj, exists, err := dswp.podInformer.GetStore().GetByKey(podKey)
+		// Check whether podUID returned from desiredStateOfWorld matches the UID retrieved from podInformer
+		// If the UIDs do not match, it means the pod in desiredStateOfWorld no longer exists, remove it
+		podExists := volumehelper.GetUniquePodName(podObj.(*api.Pod))
```

Comment:
If podObj is nil `GetUniquePodName` will choke doing pod.UID. So, first check if the cast is safe, error out if it is not. Example: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attach_detach_controller.go#L412